### PR TITLE
added basic/simple adaptive cards support

### PR DESCRIPTION
**Pull request by markush81**
### Testing done

Add starting point for AdaptiveCard Support because of Microsofts announcement.

Only manually tested in installation where i am working.

The PR is just meant as starting point to fully implement what is needed (**unfortunately** i have not the time to _add tests_ and all variants, like check _encoding of provided message_).

### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
-- https://github.com/jenkinsci/office-365-connector-plugin/issues/355
-- https://github.com/jenkinsci/office-365-connector-plugin/issues/353
-- https://github.com/jenkinsci/office-365-connector-plugin/issues/297
- [n/a] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: src/main/java/jenkins/plugins/office365connector/model/Action.java
Comment (damianszczepanik) [inline src/main/java/jenkins/plugins/office365connector/model/Action.java:5]: `Action` is very generic and very popular class name, how about making `ConnectorAction` or similar to avoid mismatch in class importing?

File: src/main/java/jenkins/plugins/office365connector/model/Action.java
Comment (damianszczepanik) [inline src/main/java/jenkins/plugins/office365connector/model/Action.java:9]: `setTargets` since this takes list as a parameter

File: src/main/java/jenkins/plugins/office365connector/model/adaptivecard/AdaptiveCard.java
Comment (damianszczepanik) [inline src/main/java/jenkins/plugins/office365connector/model/adaptivecard/AdaptiveCard.java:88]: empty method?

File: src/main/java/jenkins/plugins/office365connector/model/adaptivecard/AdaptiveCardAction.java
Comment (damianszczepanik) [inline src/main/java/jenkins/plugins/office365connector/model/adaptivecard/AdaptiveCardAction.java:5]: why full package name here?

File: src/main/java/jenkins/plugins/office365connector/model/adaptivecard/AdaptiveCardAction.java
Comment (damianszczepanik) [inline src/main/java/jenkins/plugins/office365connector/model/adaptivecard/AdaptiveCardAction.java:49]: applies only to first element, why is that ?

**Review by damianszczepanik — CHANGES_REQUESTED** [review]

File: src/main/java/jenkins/plugins/office365connector/model/Action.java
Comment (markush81) [reply src/main/java/jenkins/plugins/office365connector/model/Action.java:9]: in general i agree, but currently consistent to https://github.com/jenkinsci/office-365-connector-plugin/blob/master/src/main/java/jenkins/plugins/office365connector/model/PotentialAction.java#L57

File: src/main/java/jenkins/plugins/office365connector/model/adaptivecard/AdaptiveCard.java
Comment (markush81) [reply src/main/java/jenkins/plugins/office365connector/model/adaptivecard/AdaptiveCard.java:88]: yep, haven't used `cardThemeColor` but too much depends in the old `MessageCard` on having this method ... so i left it intentionally empty (meaning: not enough time to do bigger refactoring).

File: src/main/java/jenkins/plugins/office365connector/model/adaptivecard/AdaptiveCardAction.java
Comment (markush81) [reply src/main/java/jenkins/plugins/office365connector/model/adaptivecard/AdaptiveCardAction.java:5]: will fix it

File: src/main/java/jenkins/plugins/office365connector/model/adaptivecard/AdaptiveCardAction.java
Comment (markush81) [reply src/main/java/jenkins/plugins/office365connector/model/adaptivecard/AdaptiveCardAction.java:49]: Old `PotentialAction` works internally with list https://github.com/jenkinsci/office-365-connector-plugin/blob/master/src/main/java/jenkins/plugins/office365connector/model/PotentialAction.java#L34 i assume the `MessageCard` format expects lists. But the `AdaptiveCard`  action AFAIK doesn't. So that's why internally `url` (none-list) is used. As far as i can see, the whole code doesn't add more then one anyway. So a bigger refactoring could be, work internally with single attribute and only modify the `getTarget` in `MessageCard`  to return a list to fulfil the MessageCard JSON format.

File: src/main/java/jenkins/plugins/office365connector/model/Action.java
Comment (markush81) [reply src/main/java/jenkins/plugins/office365connector/model/Action.java:5]: Will take `CardAction` since the new way is not a `Connector` anymore, but a `PowerAutomate Workflow`.

File: src/main/java/jenkins/plugins/office365connector/model/Action.java
Comment (damianszczepanik) [reply src/main/java/jenkins/plugins/office365connector/model/Action.java]: then I believe both should be updated to be consistent

File: src/main/java/jenkins/plugins/office365connector/model/Action.java
Comment (markush81) [reply src/main/java/jenkins/plugins/office365connector/model/Action.java]: Done

File: src/main/java/jenkins/plugins/office365connector/model/Action.java
Comment (markush81) [reply src/main/java/jenkins/plugins/office365connector/model/Action.java]: Done

File: src/main/java/jenkins/plugins/office365connector/model/adaptivecard/AdaptiveCardAction.java
Comment (markush81) [reply src/main/java/jenkins/plugins/office365connector/model/adaptivecard/AdaptiveCardAction.java]: Done

**Comment by markush81** [comment]
two basic integration tests added to have at least "some" coverage.

**Comment by damianszczepanik** [comment]
What I understand from #355 with new approach we need to configure how to convert information from Jenkins into MS Teams even so I am not sure if we need to convert to adaptive card

**Comment by markush81** [comment]
> What I understand from #355 with new approach we need to configure how to convert information from Jenkins into MS Teams even so I am not sure if we need to convert to adaptive card

In can be any Card format which ideally is not deprecated or declared legacy. As far as i understand https://learn.microsoft.com/en-us/outlook/actionable-messages/message-card-reference, the current implemented format is not recommended anymore

> Microsoft recommends that new actionable message integrations use the Adaptive Card format, and existing integrations consider updating to Adaptive Card format.

Indeed the same thing says

> However, if you are sending actionable messages via an Office connector, or to a Microsoft Teams connector, you must continue to use the message card format.

Which is a) not true, from other applications we were sending Adaptive Cards since more than a year to the mentioned connectors and b) the retired the connectors.

So IMHO the best way to go forward is, to use AdaptiveCard format.

**Comment by markush81** [comment]
> Hello @markush81 , is it possible to set adaptiveCards as default? "Incoming Webhook" already has native support, it would not impact the current connector and would help in the migration to power automate.
> 
> https://learn.microsoft.com/en-us/microsoftteams/platform/webhooks-and-connectors/how-to/connectors-using?tabs=cURL%2Ctext1#send-adaptive-cards-using-an-incoming-webhook

Technically we can, but i am not sure if there is agreement for what the plugin should do in future.

**Comment by astrahan87** [comment]
@markush81 Is there any update on the status of this?

**Comment by markush81** [comment]
> @markush81 Is there any update on the status of this?

Since i am not maintainer, i am also waiting for review and merge. IMHO i have done the requested changes a while ago.

**Comment by astrahan87** [comment]
@damianszczepanik It does look like the requested changes have been made. Is there anything else?

**Comment by damianszczepanik** [comment]
One concern I have is to how to configure the hook. Based on the comments from developers I'm not even sure if we need this change as I found conclusion that with powerautomate you need to translate what is received via webhook into MS Teams message.

So concluding it would be best to update README file as well so this is clear how to configure hook with new addaptive cards and also so I'm able to test it before it gets merged.

**Comment by kikmon** [comment]
That change would be quite useful since manually converting the json in powerautomate is a bit tedious, and doesn't scale that well when you have many hooks
Having this plugin send adaptive cards will be very helpful.

**Comment by damianszczepanik** [comment]
> That change would be quite useful since manually converting the json in powerautomate is a bit tedious, and doesn't scale that well when you have many hooks Having this plugin send adaptive cards will be very helpful.

OK, this sounds like question for the answer I didn't get the answer so far!
I will review this one more time, read mentioned docs and merge this.

**Comment by marssilva** [comment]
One piece of information, adaptive cards work in the new flow with powerautomate and in the current Incoming Webhook. As a suggestion, it would be interesting to leave the default adaptive cards, aiming at the future that the current standard is discontinued.

File: src/test/resources/requests/adaptivecard_step.json
Comment (damianszczepanik) [inline src/test/resources/requests/adaptivecard_step.json:1]: name of this file does not pass exactly naming convention others

**Review by damianszczepanik — APPROVED** [review]

**Comment by damianszczepanik** [comment]
@markush81 I did review one more time. I have left a few comments but they are not blocking merging this change to master branch. This pull request is significant and valuable.
I have plan to merge #365 and then publish - take a look if you are OK with that
Also share a few details how did you test adaptive cards with this pull request and also are you planing to maintain this for a while before new version with adaptive cards support gets stable?

File: src/main/java/jenkins/plugins/office365connector/model/messagecard/MessageCard.java
Comment (markush81) [reply src/main/java/jenkins/plugins/office365connector/model/messagecard/MessageCard.java:30]: Adaptive Cards and colors are different, there is either a fixed enum which can be used, as done here `jenkins.plugins.office365connector.model.adaptivecard.AdaptiveCard#color` or do sth. like Semaphore UI does https://github.com/semaphoreui/semaphore/blob/develop/services/tasks/templates/microsoft-teams.tmpl#L44

**Comment by markush81** [comment]
> @markush81 I did review one more time. I have left a few comments but they are not blocking merging this change to master branch. 

Adjust most of the them.

> I have plan to merge #365 and then publish - take a look if you are OK with that

I am fine with the plan.

> Also share a few details how did you test adaptive cards with this pull request 

We have this version with my PR running since i created it, in our installation - so kinda live-testing.
One point of course, we use `Pipeline DSL` variant only.

> and also are you planing to maintain this for a while before new version with adaptive cards support gets stable?

I can't promise anything, but of course will do my best in case after merge issues occur.

**Comment by kikmon** [comment]
Thanks a lot to you guys :)

**Comment by damianszczepanik** [comment]
@markush81 @kikmon  version 5.0.0 has been published

**Comment by ViliusS** [comment]
I'm a little bit confused how to proceed with an upgrade to 5.0.0.
* Is there a way to send AdaptiveCard true in a Declarative Pipeline? 
* Should we switch to new Workflow app webhook on Teams side, or to Incoming Webhook connector, given that currently we are using Jenkins 365 Connector Webhook? Or Jenkins 365 Connector should work fine?

**Comment by kikmon** [comment]
As far as I know, this MR added the parameter adaptiveCards in the declarative pipeline too, so you should be good.
365 connectors will stop working next year, so you might as well start creating your workflows now.
That being said, it would be nice to have a documented 'Best Practice' use case with Workflows and adaptive cards.

**Comment by movedoa** [comment]
What do i need to change in a groovy script to send a success/unstable message.
Currently i use:

```
office365ConnectorSend webhookUrl: 'url',
            message: "${loStatus} - ${oCurNewVersion}",
            color: loColor
      }
```

As far as i gathered, color is not supported anymore, os how would i indicate succes/failure?
Also do i need to add new parameters to make the plugin send an adaptive card?

**Comment by ViliusS** [comment]
OK, I think I figured out the upgrade path:

1. If you are using **Jenkins Connector for Teams with an old webhook URL** default settings (`adaptiveCards: false`) in 5.0.0 will work . You don't need to change anything on Jenkins side.
2. If you are using **Jenkins Connector for Teams with a new improved URL** ([new URLs will be required by Microsoft as of 2025-01-01](https://learn.microsoft.com/lt-lt/microsoftteams/m365-custom-connectors#update-connectors-url)) the plugin won't work at all. I guess Jenkins Connector for Teams doesn't support new webhook URLs even though it allows to set them.
3. If you are using **Incoming Webhook Connector for Teams** then plugin should work with both `adaptiveCards: false` and `adaptiveCards: true` settings. Message formating in Teams channel will be a little bit different, but all essential information will be there.
4. If you are using **Post to a channel when a webhook request is received Power Automate Workflow** then you must set `adaptiveCards: true`. Workflows do not support other message formats.

At least this is how it behaves on my environment and Microsoft 365 tenant.

**Comment by rdrgporto** [comment]
Hi, 

An example with **Post to a channel when a webhook request is received Power Automate Workflow** would be:

```
post {
 success {
  office365ConnectorSend webhookUrl: '<webhook-url>', 
    message: 'Test with Webhook Microsoft Teams',  
    color: 'b2ff33', 
    adaptiveCards: true
 }
}
```

Regards

**Comment by damianszczepanik** [comment]
> Hi,
> 
> An example with **Post to a channel when a webhook request is received Power Automate Workflow** would be:
> 
> ```
> post {
>  success {
>   office365ConnectorSend webhookUrl: '<webhook-url>', 
>     message: 'Test with Webhook Microsoft Teams', 
>     adaptiveCards: true
>  }
> }
> ```
> 
> Regards

@rdrgporto would you prepare pull request with updated README file about adaptive cards

**Comment by damianszczepanik** [comment]
@markush81 In https://github.com/jenkinsci/office-365-connector-plugin/issues/355#issuecomment-2418896791 @ViliusS has mentioned that the plugin does not support credentials what is benefit of having adaptive cards but is still not supported by the plugin. How did you fix that problem?

**Comment by ViliusS** [comment]
@damianszczepanik maybe I was not very clear with my comments in another thread. Adaptive Card support in the plugin is working and fine. The issue I was discussing is that someone said that Workflow trigger URL is public, hence insecure. But it was also the case for Incoming Webhook or Jenkins connectors. Nothing has changed regarding security on that part.

BTW, Microsoft actually updated their documentation today: https://devblogs.microsoft.com/microsoft365dev/retirement-of-office-365-connectors-within-microsoft-teams/
The timeline for Teams/Oulook Connector deprecation is as follows:

**2025-02-01** - All connectors with old webhook URL will stop working. Users must update webhook URLs. So far as I tried, at this point, only Incoming Webhook connector is working with the updated URL. Jenkins connector doesn't work and it doesn't matter which Card format you send. I suspect it needs to be updated somehow to accept new URL format, but I doubt Microsoft will ever do that. Especially considering that that connector only accepts legacy [MessageCard](https://learn.microsoft.com/en-us/outlook/actionable-messages/message-card-reference) messages.
After this day, the only way forward is to use Workflows or general Incoming Webhook connector.
**2025 March-April** - Microsoft promises to implement MessageCard format support in Workflows.
**sometime at the end of 2025** - All Office 365 connectors are discontinued and stops working.
After that, the only working solution is to use Workflows with AdaptiveCard format as recommended, and alternative MessageCard format support if MS keeps their promises.

Thinking about AdaptiveCard support further I would vote to enable it by default now and point plugin users to upgrade instructions, or do it after 2025-02-01 which will probably mark the end of Jenkins connector on MS side anyway.

Also probably plugin naming could be revised again as AdaptiveCards are now implemented by [Cisco too](https://learn.microsoft.com/lt-lt/adaptive-cards/resources/partners). Maybe more implementations will follow.

**Comment by markush81** [comment]
> @markush81 In [#355 (comment)](https://github.com/jenkinsci/office-365-connector-plugin/issues/355#issuecomment-2418896791) @ViliusS has mentioned that the plugin does not support credentials what is benefit of having adaptive cards but is still not supported by the plugin. How did you fix that problem?

There is no authentication, but @ViliusS already commented i see.

**Comment by markush81** [comment]
> As far as i gathered, color is not supported anymore, ss how would i indicate started/success/failure?
> Also do i need to add new parameters to make the plugin send an adaptive card?

Coloring in AdaptiveCards works with a fixed set (or adding background images). So far the plugin determines the color based on build result https://github.com/jenkinsci/office-365-connector-plugin/blob/master/src/main/java/jenkins/plugins/office365connector/model/adaptivecard/AdaptiveCard.java#L40.

**Comment by rdrgporto** [comment]
> > Hi,
> > An example with **Post to a channel when a webhook request is received Power Automate Workflow** would be:
> > ```
> > post {
> >  success {
> >   office365ConnectorSend webhookUrl: '<webhook-url>', 
> >     message: 'Test with Webhook Microsoft Teams', 
> >     adaptiveCards: true
> >  }
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Regards
> 
> @rdrgporto would you prepare pull request with updated README file about adaptive cards

Hi @damianszczepanik,

It would be a example like that in [README.md](https://github.com/jenkinsci/office-365-connector-plugin/blob/master/README.md)? (after Pipeline post section):

### Pipeline post section with Adaptive Cards

```
pipeline {

    agent any

    stages {
        stage('Init') {
            steps {
                echo 'Hello!'
            }
        }
    }

    post {
        failure {
            office365ConnectorSend webhookUrl: "https://prod.westeurope.logic.azure.com:443/workflows...",
              message: 'Something went wrong', 
              status: 'Failure',
              adaptiveCards: true
        }
    }
}
````

Thanks in advance,

Regards

**Comment by damianszczepanik** [comment]
Sounds like a good example of how to use new approach - please create PR with updated README file

**Comment by aksh05** [comment]
@rdrgporto Are you planning to raise a PR with changes to README about the change. If not, would be happy to help create it. Thanks!

**Comment by rdrgporto** [comment]
Hi, @damianszczepanik and @aksh05 

I have created a Pull Request:

- https://github.com/jenkinsci/office-365-connector-plugin/pull/367

Thanks in advance,

Regards

**Pull request merged**
